datafusion/physical-expr/src/intervals/interval_aritmetic.rs

```diff
@@ -1240,6 +1270,32 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn or_test() -> Result<()> {
```

Comment:
When I was writing this test I found there appears to be something wrong with boolean interval handling. 

Specifically, `open_open(.., ..)` always creates the interval `Interval { lower: IntervalBound { value: Boolean(true), open: false }, upper: IntervalBound { value: Boolean(false), open: false } }` regardless of the arguments:

I tracked it down to the code in https://github.com/apache/arrow-datafusion/blob/4591d82ccba6f52db535a377486362a35bfd84fb/datafusion/physical-expr/src/intervals/interval_aritmetic.rs#L250-L249

And I will file a follow on ticket about this

Reply:
https://github.com/apache/arrow-datafusion/pull/7885

Reply:
When you create an interval with an open bound, if it is a lower bound, it can be `(false` or `(true`. `(false` can be represented as `[true`. For an upper bound case, `true)` rounds down to `false]`. That's why this result comes in that way.

But of course, you are right. We should not give the chance to create an invalid interval. Especially for boolean intervals, open_open case always gives an invalid interval, meaning that it needs to return an error regardless of the interval values (false or true). I will have solved all of these problems with the PR I am currently working on.

Reply:
> I will have solved all of these problems with the PR I am currently working on.

The PR you are working on sounds great 🎉 . If it is possible, can you possibly break it up into smaller parts (e.g perhaps we could break the "make `Interval` fields not `pub` in one PR, and rework how `propagate_constraints` works a a follow on PR?)